pkg/wwan/mmagent/mmdbus/client.go

```diff
@@ -0,0 +1,1500 @@
+	modemObj := c.conn.Object(MMInterface, modemPath)
+	err := c.callDBusMethod(modemObj, SignalMethodSetup, nil, c.sigPollPeriodSecs)
+	if err != nil {
+		c.log.Errorf("failed to setup signal polling period %d secs for modem %s: %v",
+			c.sigPollPeriodSecs, modemPath, err)
+	}
+}
+
+// Starts a new Go routine for modem state monitoring.
+// Client should be already locked by the caller.
+func (c *Client) startModemMonitor() {
+	c.monitorCtx, c.monitorCancel = context.WithCancel(context.Background())
+	c.monitorWG.Add(1)
+	go func() {
+		defer c.monitorWG.Done()
+		rule := fmt.Sprintf("type='signal', path_namespace='%s'", MMObjectPath)
+		c.conn.BusObject().Call(DBusMethodAddMatch, 0, rule)
+		sigChan := make(chan *dbus.Signal, dbusSigChanBuffer)
+		defer close(sigChan)
+		c.conn.Signal(sigChan)
+		defer c.conn.RemoveSignal(sigChan)
+		for {
+			select {
+			case <-c.monitorCtx.Done():
+				return
+
+			case signal := <-sigChan:
+				c.mutex.Lock()
+				c.lastMsg = time.Now()
+				c.processDbusSignal(signal)
+				c.mutex.Unlock()
+			}
+		}
+	}()
+}
+
+func (c *Client) processDbusSignal(signal *dbus.Signal) {
+	switch signal.Name {
+	// See: https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-objectmanager
+	case DBusSignalInterfacesAdded:
+		if len(signal.Body) < 2 {
+			c.log.Warnf("Unexpected body length for signal %s (%d)",
+				DBusSignalInterfacesAdded, len(signal.Body))
+			return
+		}
+		path, ok := signal.Body[0].(dbus.ObjectPath)
+		if !ok {
+			c.log.Warnf("Failed to convert path from signal %s (%v)",
+				DBusSignalInterfacesAdded, signal.Body[0])
+			return
+		}
+		if !strings.HasPrefix(string(path), ModemPathPrefix) {
+			c.log.Warnf("Skipped signal %s for path %s",
+				DBusSignalInterfacesAdded, path)
+			return
+		}
+		modem, err := c.getModem(path)
+		if err != nil {
+			c.log.Error(err)
+			return
+		}
+		prevModem, exists := c.modems[string(path)]
+		if !exists {
+			c.setupSignalPolling(path)
+			c.modems[string(path)] = modem
+			c.notifChan <- Notification{
+				Event: EventAddedModem,
+				Modem: *modem,
+			}
+		} else {
+			// Not expecting new interfaces added under a modem path
+			// other than the modem interface itself.
+			c.log.Warnf("Received signal %s for already known modem: %+v",
+				DBusSignalInterfacesAdded, signal)
+			if !prevModem.Status.Equal(modem.Status) {
+				c.notifChan <- Notification{
+					Event: EventUpdatedModemStatus,
+					Modem: *modem,
+				}
+			}
+			if prevModem.Metrics != modem.Metrics {
+				c.notifChan <- Notification{
+					Event: EventUpdatedModemMetrics,
+					Modem: *modem,
+				}
+			}
+			if prevModem.Location != modem.Location {
+				c.notifChan <- Notification{
+					Event: EventUpdatedModemLocation,
+					Modem: *modem,
+				}
+			}
+			c.modems[string(path)] = modem
+		}
+
+	// See: https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-objectmanager
+	case DBusSignalInterfacesRemoved:
```

Comment:
does it make sense to use [strategy](https://refactoring.guru/design-patterns/strategy) pattern to handle changes?

Reply:
Nope, the code is not that complex for such object oriented utopia.

Reply:
My reasoning is making it easier to parse with your eyes and narrow context of changes. I see several switch statements and wonder if it'll be easier to describe changes based on types rather then check what changes are occurring and then define logic based on type. Because with this implementation with increasing logic switch statement will grow.
I understand that we don't need it right now and this works as it is, would be glad to hear your opinion on will there be more logic around mmbus in future and weather it'll become complex enough for fab object oriented utopia with factories creating factories :D

Reply:
It shouldn't become more complex as this already covers all notifications published by ModemManager. But we will see, maybe in the future with newer modems there will be more events. In that case we might consider improving this part for more readability with some dystopian pattern magic. :)